app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileRepo.kt

```diff
@@ -0,0 +1,131 @@
+
+package org.mozilla.tv.firefox.pinnedtile
+
+import android.app.Application
+import android.arch.lifecycle.MutableLiveData
+import android.content.Context
+import android.content.SharedPreferences
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.support.annotation.AnyThread
+import android.support.annotation.UiThread
+import org.json.JSONArray
+import java.util.UUID
+
+private const val BUNDLED_SITES_ID_BLACKLIST = "blacklist"
+private const val CUSTOM_SITES_LIST = "customSitesList"
+private const val PREF_HOME_TILES = "homeTiles"
+private const val BUNDLED_HOME_TILES_DIR = "bundled"
+private const val HOME_TILES_JSON_PATH = "$BUNDLED_HOME_TILES_DIR/bundled_tiles.json"
+
+/**
+ * TODO
+ * Some methods require applicationContext in order to access /assets/
+ */
+class PinnedTileRepo(private val applicationContext: Application) {
+    private val _pinnedTiles = MutableLiveData<LinkedHashMap<String, PinnedTile>>() // FIXME: URI? String?
+    val tileCount get() = _pinnedTiles.value?.size
+
+    // TODO:  applicationContext.getSharedPreferences(PREF_HOME_TILES, Context.MODE_PRIVATE)
+    private var _sharedPreferences: SharedPreferences = applicationContext.getSharedPreferences(PREF_HOME_TILES, Context.MODE_PRIVATE)
+
+    fun loadTilesCache(): MutableLiveData<LinkedHashMap<String, PinnedTile>> {
+        val pinnedTiles = linkedMapOf<String, PinnedTile>().apply {
+            putAll(loadBundledTilesCache())
+            putAll(loadCustomTilesCache())
+        }
+
+        _pinnedTiles.value = pinnedTiles
+
+        return _pinnedTiles
+    }
+
+    fun addPinnedTile(url: String, screenshot: Bitmap?) {
+        val newPinnedTile = CustomPinnedTile(url, "custom", UUID.randomUUID()) // TODO: titles
+        _pinnedTiles.value?.put(url, newPinnedTile)
+        _pinnedTiles.value = _pinnedTiles.value
+        saveCustomTiles()
+
+        if (screenshot != null) {
+            PinnedTileScreenshotStore.saveAsync(applicationContext, newPinnedTile.id, screenshot)
+        }
+    }
+
+    /**
+     * returns tile id of a Bundled tile or null if
+     * it doesn't exist in the cache
+     */
+    @UiThread
+    fun removePinnedTile(url: String): String? {
+        val tileToRemove = _pinnedTiles.value?.remove(url) ?: return null
+        _pinnedTiles.value = _pinnedTiles.value
+
+        when (tileToRemove) {
+            is BundledPinnedTile -> {
+                val blackList = loadBlacklist()
+                blackList.add(tileToRemove.id)
+                saveBlackList(blackList)
+            }
+            is CustomPinnedTile -> {
+                saveCustomTiles()
+                PinnedTileScreenshotStore.removeAsync(applicationContext, tileToRemove.id)
+            }
+        }
+
+        return tileToRemove.idToString()
+    }
+
+    @UiThread
+    fun isUrlPinned(url: String): Boolean? = _pinnedTiles.value?.containsKey(url)
+
+    private fun loadBlacklist(): MutableSet<String> {
```

Comment:
I don't think there's any reason to make these mutable: who knows what mutating a set from shared prefs would do!

Reply:
Actually, this should be mutable because `blackList.add(tileToRemove.id)` when removing a bundled pinned tile